rfd/0127-encrypted-session-recordings.md

```diff
@@ -0,0 +1,489 @@
+service ClusterConfigService {
+  // existing messages omitted
+
+  // EncryptedSessionRecordingChunk is an individual chunk of an encrypted 
+  // session recording .tar file.
+  message EncryptedSessionRecordingChunk {
+    // SessionID the recording relates to.
+    string SessionID = 1;
+    // ChunkIndex is the ordered index applied to the chunk.
+    int64 ChunkIndex = 2;
+    // Chunk is the encrypted chunk of session recording data being uploaded.
+    bytes Chunk = 3;
+  }
+
+  message UploadEncryptedSessionRecordingResponse {}
+
+  service AuthService {
+    // existing RPCs omitted
+
+    // UploadEncryptedSessionRecording is used to upload encrypted .tar files
+    // containing session recording events into long term storage.
+    rpc UploadEncryptedSessionRecording(stream EncryptedSessionRecordingChunk) returns (UploadEncryptedSessionRecordingResponse);
+    
+  }
+}
+```
+
+### Session Recording Modes
+
+There are four session recording modes that describe where the recordings are
+captured and how they're shipped to long term storage.
+- `proxy-sync`
+- `proxy`
+- `node-sync`
+- `node`
+Where the recordings are collected is largely unimportant to the encryption
+strategy, but whether or not they are handled async or sync has different
+requirements.
+
+In sync modes the session recording data is written immediately to the auth
+service without intermediate disk writes. The auth service then handles
+batching and multipart upload of events to long term storage. In order to
+ensure resilience to partial uploads of parts, each batch will need to be
+encrypted individually on the auth service. This will likely mean wrapping the
+`gzipWriter` [used by slices](https://github.com/gravitational/teleport/blob/master/lib/events/stream.go#L669)
+with an `age` encryption writer.
+
+In async modes the session recording data is written to intermediate `.part`
+files. These files are collected until they're ready for upload and are then
+combined into a single `.tar` file. In order to encrypt individual parts, we
+will build a special `io.Writer` that contains a single instance of the age
+encryptor that can proxy writes across multiple files. This will require
+maintaining a concurrency-safe mapping of in-progress uploads to encrypted
+writers which incurs a bit of added complexity. However it intentionally avoids
+any sort of intermediate key management which seems a worthwhile tradeoff.
+Because the data is already encrypted, exploding the final `.tar` file back
+into events to be uploaded to the auth service is not possible. Instead the
+auth service must accept a binary upload of the `.tar` file which it can
+then proxy to long term storage. This will be done using the
+`UploadEncryptedSessionRecording` streaming RPC. When using S3 compatible long
+term storage, the auth service will need to use the `PutObject` API rather than
+performing a multipart upload. This is to prevent any loss of parts that would
+make decryption impossible.
```

Comment:
i assume there's a reason we currently do multi-part uploads, could sessions be too large for a single PutObject? just a question idk

Reply:
The maximum size for `PutObject` is 5GB, so I wouldn't think it's super likely? As far as I can tell we do the multi-part upload so that we aren't holding onto large amounts of recording data in memory while we wait for the session to conclude. The issue of the agent crashing halfway through definitely needs to be addressed, though. I think it would force us to encrypt each part/chunk individually similar to how `sync` modes would work. Even so, we would need this new RPC to upload the encrypted chunks since we can't deconstruct them back into events at the node/proxy level.